### ISSUE TITLE
[Grouping] Custom aggregation functions

### ISSUE THREAD

**Issue by begemott** [angular-ui/ng-grid#3302]
Hi

I've tried the new grouping feature and find out that it is impossible to provide a function that calculates a group value. For example, if I want to group by a column that contains DateTime values, now the Grid creates a group for each distinct value in the column. Instead it would be better to have groups like Today, Yesterday, This Week, This Month and so on. It can be done by adding a new field like valueFormatter or valueBuilder in the columnDef.grouping object. The type of the field is function that accept a cell value and return a formatted group value.

**Comment by c0bra** [comment]
Thanks for the suggestion!

**Comment by AgDude** [comment]
I am looking into what it will take to implement this. Note that the current implementation holds a "running" value as we iterate over the rows, which will make writing these functions a little less intuitive than simply operating on an array.

For reference, this is the list of functions I need to implement for a project I am migrating to uiGrid:

* 'average'
* 'mean'
* 'median'
* 'mode'
* 'stdevp'
* 'varp'
* 'stdev'
* 'var'
* 'min'
* 'max'
* 'sum'
* 'count'
* 'list' - In our application, this concatenates unique values.
* 'concat' - concatenate all values (may need to set a length limit)

**Comment by AgDude** [comment]
Here is my current thought on an API. I haven't started implementing this, but I am interested in feedback on whether this sounds reasonable. One of the challenges I am considering is calculating something like "median" or "mode", where you need to have access to the entire array (admittedly at a performance cost).

Add a columnDef option ```groupingAggregationTypes``` which is an object where the keys are the aggregation name, and the values could be either a function or an object with two functions. The first function would need to accept the ```aggregation``` object which is stored in ```processingState.runningAggregations``` and the ```fieldValue```, and would be responsible for modifying the ```value``` property of the object. It could also add/modify other properties if needed. Providing two functions would allow you to use one function to build the ```aggregation``` object during iteration, and another to be executed at ```writeOut```. The ```writeOut``` method of the service should probably do some cleanup also, so when it re-runs it will always start with a fresh ```aggregation``` object.

For example this would be the option object to add 'mean' (same as build in average) and the 'median':

    {mean: function(aggregation, fieldValue){
        aggregation.count++;
        if (!isNaN(numValue)){
          aggregation.sum += fieldValue;
         }
         aggregation.value = aggregation.sum / aggregation.count;
      },
    median: {
      running: function(aggregation, fieldValue){
        if ( typeof aggregation.allValues === 'undefined' ) { aggregation.allValues = []; }
        aggregation.allValues.push(fieldValue);
      },
      writeOut: function(aggregation){
         aggregation.allValues.sort();
         var half = Math.floor(aggregation.allValues.length/2);
         return aggregation.allValues[half];
      }
    }

This proposal is by no means simple, so if you have other ideas please share. I think it is unavoidable that we sometimes need a way to process the result after all rows have been read. One option is that custom functions will always be presented with all of the values, and simply take a performance hit compared to the built-in ones.

**Comment by AgDude** [comment]
Another possible way to handle functions which need all of the data would be to support only running functions, but add an event ```beforeWriteOutAggregation```, which would allow the value to be modified. This probably makes implementing those functions even more complicated, but I think it would take some of the responsibility out of uiGrid and put it on the user. Another advantage to this method would be it would allow modifying of a built-in function (for example rounding an avg).

**Comment by begemott** [comment]
I think the first option is better.

**Comment by PaulL1** [comment]
This should be included in my treeBase branch.

**Comment by PaulL1** [comment]
This should be merged now, can you take a look when you get a chance?

**Comment by AgDude** [comment]
@PaulL1 I am looking at this today. This looks like  it should accomplish what I had in mint, but I have some thoughts on cleaning up the api a bit. If you are working on grid issues today, try to catch me on gitter. I am happy to write the code, but want to run some thoughts past you first. I like the two grid options of ```customTreeAggregationFn``` and ```customTreeAggregationFinalizerFn``` (especially the fact that the finalizer can be used with builtin functions.)

Here are a few things on my mind:
* Let's make it easier to put a custom aggregations in the menu (probably a public api method)
* Refactor the builtin aggregators to be more similar to custom aggregators. This would allow the user to still choose a builtin aggregator with a string, but then we assign a function to the ```col``` object, and get rid of the switch/case in treeBase.aggregate. I think it would remove a few if/else statements in other places too.

**Comment by AgDude** [comment]
Here is a plunker which demonstrates a median calculation and a couple flexible factories for creating custom aggregators. Maybe I'll add this to the tutorial at some point.

Be warned that this adds an additional sort within each group, which will be slow on large datasets.
http://plnkr.co/edit/3lAuYzzGBwaQ3PJRSvh8?p=preview

**Comment by PaulL1** [comment]
I'm OK with the concepts you describe:
 - yes to adding custom aggregations to the menu somehow
 - yes to setting the function on the col rather than using the switch each time
 - hmm.  if you have a way to apply the built-in finalizer then you'll probably end up with an option and messing around, all to save one line of code.....

**Comment by AgDude** [comment]
On the last one I was thinking something along the lines of if ```aggregation.rendered``` is undefined. Since basically all it does is copy the value to ```rendered``` and apply the label. We'll also need to check for a label, which the user will need a way to provide anyway if it is going in the menu.

**Comment by PaulL1** [comment]
That's probably fair.

**Comment by AgDude** [comment]
@PaulL1 Do you have any objection to changing the columnDef option of ```treeAggregation``` into two options ```treeAggregationType``` and ```treeAggregationLabel```? It just seems inconsistent to make the custom functions separate options and the others an object.  Alternatively I could put the custom functions inside the object but I think the documentation is clearer when they are separate options.

I would leave the old option as well for the time being. Not sure if we have a clear deprecation path.

**Comment by PaulL1** [comment]
That's fine.  I wouldn't bother deprecating - nobody has had time to use it yet to be worth it (this is all on the new treeAggregation option anyway).  I didn't bother deprecating all the large changes I did in the API on the basis that we haven't pushed 3.0 yet, and it would have been quite untidy.  Grouping is still marked alpha.....

**Comment by AgDude** [comment]
I have most of the discussed changes done, but I am now thinking about another API change, which would make custom functions easier to work with in saveState. Instead of specifying the ```aggregationFn``` on each column, there would be a grid level option ```treeAggregationFns```, which is an object like 
    
    { standardDeviation: {label: 'stdDev', aggregationFn: Function, finalizerFn: Function},
      variance: {label: 'var', aggregationFn: Function, finalizerFn: Function},
     }

Each column can then get an ```aggregationType``` as a string which is either one of the native options, or one defined at the grid level.

**Comment by PaulL1** [comment]
Hmm.  Doesn't grab me, but I see the logic.  And has the advantage of not repeating the functions.  I was trying to reach a point where the functions also didn't get repeated on each row, which I can't recall whether I achieved or not.  Not having them in the aggregation entity was key for that - I had tried to allow for there to be custom values in the aggregation setting that I'd copy blindly without knowing about them - so people could have other settings in their customisation.  Just something to watch out for.
